operators/pkg/exmagnt/instance.go

```diff
@@ -0,0 +1,270 @@
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(w, "Something went wrong. Please retry later")
+		klog.Errorf("Instance %s INVALID PHSE: %v", inst.Name, inst.Status.Phase)
+		return
+	}
+
+	klog.Infof("Instance %s (phase: %v): sending starting-up page", inst.Name, inst.Status.Phase)
+
+	w.Header().Add("refresh", "5")
+	w.WriteHeader(http.StatusCreated)
+	fmt.Fprint(w, httpPageStartingUp)
+}
+
+// HandlePut handles the PUT request for a EAInstance api call.
+func (i *EAInstanceHandler) HandlePut(w http.ResponseWriter, r *http.Request) {
+	if !Options.CheckAllowedIP(r) {
+		klog.Errorf("Request from unauthorized IP: %v", r.RemoteAddr)
+		w.WriteHeader(http.StatusForbidden)
+		fmt.Fprint(w, "Forbidden")
+		return
+	}
+
+	// get Instance from the request
+	eaInstance, err := EAInstanceFromRequest(r)
+	if err != nil {
+		klog.Errorf("Error parsing request: %v", err)
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprint(w, "Bad request")
+		return
+	}
+
+	inst := EmptyInstanceFromRequest(r)
+	instSpec := InstanceSpecFromEAInstance(&eaInstance)
+
+	op, err := ctrl.CreateOrUpdate(r.Context(), Client, inst, func() error {
+		inst.Spec = instSpec
+		return nil
+	})
+
+	if err != nil {
+		klog.Errorf("Instance %s cannot be %s: %v", inst.Name, op, err)
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "Instance %s cannot be %s", inst.Name, op)
+		return
+	}
+
+	w.Header().Add("Content-Type", "text/plain")
+
+	switch op {
+	case ctrlutil.OperationResultCreated:
+		w.WriteHeader(http.StatusCreated)
+	case ctrlutil.OperationResultUpdated:
+		w.WriteHeader(http.StatusOK)
+	default:
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+
+	fmt.Fprint(w, op)
+	klog.Errorf("Instance %s has been %s", inst.Name, op)
+}
+
+// HandleGetAll handles the GET request for all the instances.
+func HandleGetAll(w http.ResponseWriter, r *http.Request) {
+	var instances clv1alpha2.InstanceList
+	if err := Client.List(r.Context(), &instances); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(w, "Error retrieving instances")
+		klog.Errorf("Error retrieving instances: %v", err)
+		return
+	}
+
+	eaInstances := make([]EAInstance, len(instances.Items))
+	for i := range instances.Items {
+		eaInstances[i] = *EAInstanceFromInstance(&instances.Items[i])
+	}
+
+	w.Header().Add("Content-Type", "application/json")
+	if err := json.NewEncoder(w).Encode(eaInstances); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(w, "Internal server error")
+		klog.Errorf("Error encoding instances: %v", err)
+		return
+	}
+}
+
+// GetInstanceIDFromRequest returns the instance id from the request.
+func GetInstanceIDFromRequest(r *http.Request) string {
+	InstanceEP := path.Join(Options.BasePath, EAInstanceEndpoint) + "/"
+	instID := strings.Replace(r.URL.Path, InstanceEP, "", 1)
+	return instID
```

Comment:
Hoped there was a better way

Reply:
can think of any?

Reply:
None actually. Would just save InstanceEP as field of the handler structure, to avoid having it hardcoded here (wouldn't allow to support two paths for instance)